crates/handlers/src/compat/login.rs

```diff
@@ -33,19 +33,32 @@ use thiserror::Error;
 
 use super::MatrixError;
 
+#[derive(Debug, Serialize)]
+enum Action {
+    #[serde(rename = "login")]
+    Login,
+    #[serde(rename = "register")]
+    Register,
+}
+
 #[derive(Debug, Serialize)]
 #[serde(tag = "type")]
 enum LoginType {
     #[serde(rename = "m.login.password")]
-    Password,
+    Password {
+        actions: Vec<Action>,
+    },
 
     #[serde(rename = "m.login.token")]
-    Token,
+    Token {
+        actions: Vec<Action>,
```

Comment:
Should this one have be empty array? If not, I'd either remove it here, or add
```
#[serde(skip_serializing_if = "Vec::is_empty")]
```
so it skips the field if the array is empty

Reply:
As per discussion on Matrix, I've removed `actions` for `m.login.token`.